docs/source/developers/guide/step_by_step/building.rst

```diff
@@ -35,17 +35,83 @@
 Building Arrow's libraries 🏋🏿‍♀️
 *********************************
 
+If you decide to contribute to Arrow you will meet the topic of
+compiling source code and use of CMake. You may have some
+experience with it or not. If not, it is good to read through
+this part so you understand what is happening in the process of
+building Arrow better.
 
+If you feel comfortable with compiling then feel free to proceed
+to the :ref:`C++ <building-arrow-cpp>`, :ref:`PyArrow <build_pyarrow>` or
+`R package build section <https://arrow.apache.org/docs/r/articles/developing.html>`_.
 
 Building C++
 ============
 
+Why build C++ from source?
+--------------------------
+
+The core of Arrow is written in C++ and all bindings in other
+languages (Python, R, ..) are wrapping underlying
+C++ functions. Even if you want to work on PyArrow or R package
```

Comment:
This is only true for the C++ implementation and its bindings. We should maybe try to reword this a bit to more correctly represent the fact that there are non-C++ Arrow implementations